Design/spectrum_number_and_workspace_index_abstraction.md

```diff
@@ -0,0 +1,212 @@
+- There is no consistent way of defining index ranges or lists for algorithms.
+  This is done individually in each algorithm, and properties may be defined for spectrum numbers or workspace indices, and sometimes the interpretation of index properties even seems to depend on the facility (`LoadEventNexus`, see [13475](https://github.com/mantidproject/mantid/issues/13475)).
+- Validation of index ranges and behavior on errors such as out-of-range indices is likewise done individually in all algorithms.
+  - Algorithms behave inconsistently depending on errors (throw and error, print a warning, ignore silently).
+  - There are bugs in some range validations (see [15414](https://github.com/mantidproject/mantid/issues/15414)).
+  - Some algorithms do not validate for duplicate spectrum numbers and double-process the corresponding spectra (see [16651](https://github.com/mantidproject/mantid/issues/16651)).
+  - Validations are not unit-tested to a good extent.
+
+#### Future problems:
+
+When running Mantid with MPI we will store different spectra on different MPI ranks.
+As a consequence the workspace index (as we would get it as user-input via an algorithm property) will no longer correspond to the memory offset in the local vector of spectra.
+Basically we will have a global index and a local index on each rank, as exemplified above.
+The local indices from all ranks are disjoint sets that together make up all global indices.
+
+This introduces additional complexity in the index handling.
+It is pretty clear that we should not and cannot put the burden of dealing with the global indexing scheme on all our (algorithm) developers.
+A lot of care needs to be taken when translating between local and global indices since any mistake will lead to incorrect results that can be very hard notice and are very difficult to trace down.
+It is thus absolutely crucial that this is done only in a single location in our code base and is unit tested well.
+
+#### Instrument 2.0:
+
+Instrument 2.0 currently under [prototype](https://github.com/DMSC-Instrument-Data/instrument-prototype) will come with MPI support as per the [design](https://github.com/DMSC-Instrument-Data/documents/tree/master/meeting_notes/April_2016/InstrumentPrototype#mpi-support).
+Cleaning up our index handling is thus a prerequisite for its implementation.
+
+
+## Detailed goals
+
+- All algorithms should have a consistent interface.
+- Eliminate duplicated code for declaring properties for index ranges and list.
+- Eliminate duplicated code for validating indices and converting them to the desired type.
+- Developers should generally not have to deal with spectrum number or detector ID at all.
+  Spectrum number and detector ID are important for the user interface (or file interface), but irrelevant for the internals of most algorithms.
+- Developers must be isolated from dealing explicitly with the local and global indices that come with an MPI implementation of Mantid.
+
```

Comment:
Splitting into several properties (as you describe below) suggests that the inputs will be qualified to be either SpectrumNumber/GlobalIndex, or DetectorIndexes. If you haven't discussed this with @NickDraper you might want to do so, because I recall him talking about adding some kind of universal translation service, although that might have been at the UI level only (i.e. No changes to the existing `FloatArrayProperty` properties of any algorithm.